[ext-doc] New SidePanel API documentation (#6178)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/reference/sidePanel/index.md

---
api: sidePanel
---

## Overview {: #overview }

> Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The SidePanel API allows extensions to add their own panels to the Chrome browser. Some features include:

Reviewer: Can we add a screenshot of a side panel here?

Reviewer: I like the first one! It takes up less space and also shows where to open it from.

> Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The SidePanel API allows extensions to add their own panels to the Chrome browser. Some features include:

Reviewer: ```suggestion
Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The Side Panel API allows extensions to display their own UI in the side panel. Some features include:
```

Author: DOne 👍

- The side panel remains open when navigating between tabs (if set to do so).
- It can be available only on specific websites.
- As an extension page, side panels have access to all Chrome APIs.

> - Users can choose which side to display the panel under Chrome settings.

Reviewer: ```suggestion
- Users can choose which side to display the panel on within Chrome's settings.
```

## Manifest {: #manifest }

To use the SidePanel API, add the `"sidePanel"` permission in the extension [manifest][doc-manifest] file:

{% Label %}manifest.json:{% endLabel %}

```json
{
  "name": "My side panel extension",
  ...
  "permissions": [
    "sidePanel"
  ]
}
```

## Use cases {: #use-cases }

The following sections demonstrate some common use cases for the SidePanel API. See [Extension samples](#examples) for complete extension examples.

### Display the same side panel on every site {: #every-site }

> The side panel can be set initially from the `“default_path”` property in the `“side_panel”` key of the manifest to display the same side panel on every site. This should point to a relative path within the extension directory.

Reviewer: I think something might've messed up your quotes here?

```suggestion
The side panel can be set initially from the `"default_path"` property in the `"side_panel"` key of the manifest to display the same side panel on every site. This should point to a relative path within the extension directory.
```

Author: Gah 🫢, thanks for catching that!

{% Label %}manifest.json:{% endLabel %}

```json
{
  "name": "My side panel extension",
  ...
  "side_panel": {
    "default_path": "sidepanel.html"
  }
  ...
}
```

{% Label %}sidepanel.html:{% endLabel %}

```html
<!DOCTYPE html>
<html>
  <head>
    <title>My Sidepanel</title>
  </head>
  <body>
    <h1>All sites sidepanel extension</h1>
    <p>This side panel is enabled on all sites</p>
  </body>
</html>
```

### Enable a side panel on a specific site {: #by-site }

> An extension can use [`sidepanel.setOptions()`][sidepanel-setoptions] to enable a side panel on a specific site. This example uses [`chrome.tabs.onUpdated()`][tabs-onupdated] to listen for any updates made to the tab. It checks if the URL is [www.google.com](https://www.google.com) and enables the side panel. Otherwise, it disables it. 

Reviewer: I think it makes sense to say what the API can do (which is set it for a tab) and then in the later sentence we talk about what that gives you?

```suggestion
An extension can use [`sidepanel.setOptions()`][sidepanel-setoptions] to enable a side panel on a specific tab. This example uses [`chrome.tabs.onUpdated()`][tabs-onupdated] to listen for any updates made to the tab. It checks if the URL is [www.google.com](https://www.google.com) and enables the side panel. Otherwise, it disables it. 
```

{% Label %}service-worker.js:{% endLabel %}

```js
const googleURL = 'https://www.google.com';

chrome.tabs.onUpdated.addListener((tabId, info, tab) => {
  if (!tab.url) return;
  const url = new URL(tab.url);
  // Enables the side panel on google.com
  if (url.origin === googleURL) {
    chrome.sidePanel.setOptions({
      tabId,
      path: 'sidepanel.html',
      enabled: true
    });
  // Disables the side panel on all other sites
  } else {
    chrome.sidePanel.setOptions({
      tabId,
      enabled: false
    });
  }
});
```

> When the user navigates to a site where the side panel is not enabled, the side panel will close and not be available on the side panel drop-down menu.

Reviewer: ```suggestion
When a user temporarily switches to a tab where the side panel is not enabled, the side panel will be hidden. It will automatically show again when the user switches to a tab where it was previously open.

When the user navigates to a site where the side panel is not enabled, the side panel will close and the extension will not show in the side panel drop-down menu.
```

> For a complete example, see the[Tab-specific side panel][sample-sp-google] sample. 

Reviewer: ```suggestion
For a complete example, see the [Tab-specific side panel][sample-sp-google] sample. 
```

---
api: sidePanel
---

{% Aside 'important' %}

> The Side Panel API will launch on [Chrome 114](https://chromiumdash.appspot.com/schedule), available in stable in late May 2023

Reviewer: ```suggestion
The Side Panel API will launch in [Chrome 114](https://chromiumdash.appspot.com/schedule), currently in beta and available in stable in late May 2023.
```

      enabled: false
    });
  }
});
```

When a user temporarily switches to a tab where the side panel is not enabled, the side panel will be hidden. It will automatically show again when the user switches to a tab where it was previously open.

When the user navigates to a site where the side panel is not enabled, the side panel will close and the extension will not show in the side panel drop-down menu.

For a complete example, see the [Tab-specific side panel][sample-sp-google] sample. 

### Enable the action icon to open the side panel {: #open-action-icon } 

Developers can allow users to open the side panel when they click on the action toolbar icon with [`sidePanel.setPanelBehavior()`][sidepanel-set-behavior]. First, declare the `"action"` key in the manifest:

{% Label %}manifest.json:{% endLabel %}

```json/3-5
{
  "name": "My side panel extension",
  ...
   "action": {
    "default_title": "Click to open panel"
  },
  ...
}
```

Now, let's add this functionality to the previous example:

{% Label %}service-worker.js:{% endLabel %}

```js/3-3
const GOOGLE_ORIGIN = 'https://www.google.com';

// Allows users to open the side panel by clicking on the action toolbar icon

> chrome.sidePanel.setPanelBehavior({ openPanelOnActionClick: true });

Reviewer 2: Same would apply to `setOptions` in other examples

> chrome.sidePanel.setPanelBehavior({ openPanelOnActionClick: true });

Author: I appreciate your feedback @beaufortfrancois  😄 

I decided to catch the error on the first call and add `async` to `setOptions`.

---
api: sidePanel
---

{% Aside 'important' %}
The Side Panel API will launch in [Chrome 114](https://chromiumdash.appspot.com/schedule), currently in beta and available in stable in late May 2023.
{% endAside %}

## Overview {: #overview }

> Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The Side Panel API allows extensions to display their own UI in the side panel. 

Reviewer 3: Would suggest: "... their own UI in the side panel, enabling persistent experiences that complement the user's browsing journey."

Author: Love it! Let's do that 👍 💯

> Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The Side Panel API allows extensions to display their own UI in the side panel. 

Author: ```suggestion
Chrome features a built-in side panel that enables users to view more information alongside the main content of a webpage. The Side Panel API allows extensions to display their own UI in the side panel, enabling persistent experiences that complement the user's browsing journey. 
```

<figure>
  {% Img src="image/BhuKGJaIeLNPW9ehns59NfwqKxF2/2uFG8qxM7cqyMuXWlD9R.png", alt="Side panel drop-down menu", width="291", height="366" %}
  <figcaption>
    Chrome browser side panel UI.
  </figcaption>
</figure>

Some features include:

- The side panel remains open when navigating between tabs (if set to do so).
- It can be available only on specific websites.
- As an extension page, side panels have access to all Chrome APIs.

> - Users can choose which side to display the panel on within Chrome's settings.

Author: ```suggestion
- Within Chrome's settings, users can specify which side the panel should be displayed on.
```
